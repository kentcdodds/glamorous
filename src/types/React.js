// @flow
import type {Component, Element} from 'react'

export type FunctionalComponent = (
  {[key: string]: any}, // props
  ?{[key: string]: any}, // context
) => Element<*> | null;

// eslint-disable-next-line no-undef
export type ComponentClass = Class<Component<*, *, *>>;

export type RefCallback<Instance> = (Instance | null) => void;
