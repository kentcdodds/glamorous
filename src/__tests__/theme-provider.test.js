/* eslint func-style:0 */
import React, {Component} from 'react'
import {render, mount} from 'enzyme'
import * as jestGlamorReact from 'jest-glamor-react'
import toJson from 'enzyme-to-json'
// eslint-disable-next-line import/default
import glamorous from '../'
import ThemeProvider, {CHANNEL} from '../theme-provider'

expect.extend(jestGlamorReact.matcher)
expect.addSnapshotSerializer(jestGlamorReact.serializer)

const getMockedContext = unsubscribe => ({
  [CHANNEL]: {
    getState: () => {},
    setState: () => {},
    subscribe: () => unsubscribe,
  },
})

test('renders a component with theme', () => {
  const Comp = glamorous.div(
    {
      color: 'red',
    },
    (props, theme) => ({padding: theme.padding}),
  )
  expect(
    render(
      <ThemeProvider theme={{padding: '10px'}}>
        <Comp />
      </ThemeProvider>,
    ),
  ).toMatchSnapshotWithGlamor()
})

test('theme properties updates get propagated down the tree', () => {
  class Parent extends Component {
    state = {
      padding: '10px',
    }
    setPadding = () => this.setState({padding: '20px'})
    render() {
      return (
        <ThemeProvider theme={{padding: this.state.padding}}>
          <Child />
        </ThemeProvider>
      )
    }
  }
  const Child = glamorous.div(
    {
      color: 'red',
    },
    (props, theme) => ({padding: theme.padding}),
  )
  const wrapper = mount(<Parent />)
  expect(toJson(wrapper)).toMatchSnapshotWithGlamor(`enzyme.mount`)
  wrapper.instance().setPadding()
  expect(toJson(wrapper)).toMatchSnapshotWithGlamor(`enzyme.mount`)
})

test('merges nested themes', () => {
  const One = glamorous.div({}, (props, themes) => ({padding: themes.padding}))
  const Two = glamorous.div({}, (props, themes) => {
    return {padding: themes.padding, margin: themes.margin}
  })
  expect(
    mount(
      <div>
        <ThemeProvider theme={{padding: '1px'}}>
          <div>
            <One />
            <ThemeProvider theme={{margin: '2px'}}>
              <Two />
            </ThemeProvider>
          </div>
        </ThemeProvider>
      </div>,
    ),
  ).toMatchSnapshotWithGlamor()
})

test('renders if children are null', () => {
  expect(
    mount(
      <ThemeProvider theme={{padding: '1px'}}>
        {false && <div />}
      </ThemeProvider>,
    ),
  ).toMatchSnapshotWithGlamor()
})

test('cleans up outer theme subscription when unmounts', () => {
  const unsubscribe = jest.fn()
  const context = getMockedContext(unsubscribe)
  const wrapper = mount(<ThemeProvider theme={{}} />, {context})
  wrapper.unmount()
  expect(unsubscribe).toHaveBeenCalled()
})

test('does nothing when receive same theme via props', () => {
  const theme = {margin: '2px'}
  const wrapper = mount(<ThemeProvider theme={theme} />)
  expect(toJson(wrapper)).toMatchSnapshotWithGlamor(`enzyme.mount`)
  wrapper.setProps({theme})
  expect(toJson(wrapper)).toMatchSnapshotWithGlamor(`enzyme.mount`)
})
