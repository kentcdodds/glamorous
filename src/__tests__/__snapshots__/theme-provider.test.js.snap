// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enzyme.mount 1`] = `
.css-1fyph4i,
[data-css-1fyph4i] {
  color: red;
  padding: 10px;
}

<Parent>
  <ThemeProvider
    theme={
      Object {
        "padding": "10px",
      }
    }
  >
    <glamorous(div)>
      <div
        className="css-1fyph4i"
      />
    </glamorous(div)>
  </ThemeProvider>
</Parent>
`;

exports[`enzyme.mount 2`] = `
.css-1t91z0n,
[data-css-1t91z0n] {
  color: red;
  padding: 20px;
}

<Parent>
  <ThemeProvider
    theme={
      Object {
        "padding": "20px",
      }
    }
  >
    <glamorous(div)>
      <div
        className="css-1t91z0n"
      />
    </glamorous(div)>
  </ThemeProvider>
</Parent>
`;

exports[`enzyme.mount 3`] = `
<ThemeProvider
  theme={
    Object {
      "margin": "2px",
    }
  }
/>
`;

exports[`enzyme.mount 4`] = `
<ThemeProvider
  theme={
    Object {
      "margin": "2px",
    }
  }
/>
`;

exports[`merges nested themes 1`] = `
.css-e5hgul,
[data-css-e5hgul] {
  padding: 1px;
}

.css-45n9js,
[data-css-45n9js] {
  padding: 1px;
  margin: 2px;
}

<div>
  <ThemeProvider
    theme={
      Object {
        "padding": "1px",
      }
    }
  >
    <div>
      <glamorous(div)>
        <div
          className="css-e5hgul"
        />
      </glamorous(div)>
      <ThemeProvider
        theme={
          Object {
            "margin": "2px",
          }
        }
      >
        <glamorous(div)>
          <div
            className="css-45n9js"
          />
        </glamorous(div)>
      </ThemeProvider>
    </div>
  </ThemeProvider>
</div>
`;

exports[`renders a component with theme 1`] = `
.css-1fyph4i,
[data-css-1fyph4i] {
  color: red;
  padding: 10px;
}

<div
  class="css-1fyph4i"
/>
`;

exports[`renders if children are null 1`] = `
<ThemeProvider
  theme={
    Object {
      "padding": "1px",
    }
  }
/>
`;
